[Security Solution][Detections]Rules _bulk_action edit API (#1317)
elastic/security-docs: docs/detections/api/rules/rules-api-bulk-actions.asciidoc

[[bulk-actions-rules-api-action]]

> ==== Bulk action

Reviewer: This is an existing page question, but any reason `Create`/`Update` have the `this API supports Token-based authentication` warning 
<p align="center">
  <img width="350" src="https://user-images.githubusercontent.com/2946766/156076269-6193bcb9-05a9-455d-a835-44586c2c91d0.png" /> <img width="350" src="https://user-images.githubusercontent.com/2946766/156076382-dda37bfd-9fbe-472b-ab06-1e8ee94b6fd1.png" />
</p>

but `Delete`/`Action` do not?
<p align="center">
  <img width="350" src="https://user-images.githubusercontent.com/2946766/156076452-b1948406-1d66-40e8-8aa4-719709c60f8e.png" /> <img width="350" src="https://user-images.githubusercontent.com/2946766/156076536-92a9c5a2-135b-4109-99ed-db20a1413d82.png" />
</p>

It's the same for the whole API, so can this just be a page-level warning?

Author: @spong, not sure about this. Is there a way to determine which type of authentication is supported by particular API(maybe looking into route configuration)? 

Take for example this one, read rule https://www.elastic.co/guide/en/security/current/rules-api-get.html
and create rule https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all

Create API has warning, but by looking into route configuration I don't see the difference. We might want to align these warnings to be consistent across different APIs

Reviewer: > @spong, not sure about this. Is there a way to determine which type of authentication is supported by particular API(maybe looking into route configuration)?

I believe so long as security is enabled (which it is by default now) all routes will go through this auth ([source](https://github.com/elastic/kibana/blob/6627bd8b3ab49415e397495d8e076f9861fcc894/x-pack/plugins/security/server/authentication/providers/token.ts#L101-L106)).

So yeah, I think it's just a consistency thing -- generally, auth shouldn't vary between these Kibana CRUD routes, so I think we just need to align on these warnings like you said.

> Applies a bulk action to multiple rules. The bulk action is applied to all rules that match the filter or to the list of rules by their ids.

Reviewer 2: ```suggestion
Applies a bulk action to multiple rules. The bulk action is applied to all rules that match the filter or to the list of rules by their IDs.
```

[discrete]
===== Request body

A JSON object with the following properties:

[width="100%",options="header"]
|==============================================
|Name |Type |Description |Required
| `query` | String |  A string containing a KQL search query to match the rules. | No
| `ids` | String[] |  Array of rule ids, to which bulk action will be applied. Only valid when query property is undefined | No
| `action` | Enum a| A bulk action
to apply

.Possible values
* `enable`
* `disable`
* `delete`
* `duplicate`
* `export`
* `edit`

| Yes
| `edit` | <<bulk-edit-object-schema, BulkEditAction[]>>
| Edit object that describes applying an update action

|No.

Yes, if action is `edit`
|==============================================

[[bulk-edit-object-schema]]
[discrete]
==== BulkEditAction object

> * `type` - enum - that defines what will be updated in rules

Author: <img width="676" alt="Screenshot 2022-03-03 at 12 38 40" src="https://user-images.githubusercontent.com/92328789/156566343-1a430217-6048-42fb-bc64-6466f188b74c.png">

this enum is already incorporated in table

* `value` - any - value which will be applied in edit action.

[discrete]
===== Possible `BulkEditAction` object values

[width="100%",options="header"]
|==============================================
|`type` field |`value` field |Description
| `add_tags` | String[] |  Adds tags to rules
| `delete_tags` | String[] |  Delete tags from rules

> | `set_tags` | String[] |  Set value as tags in rules

Reviewer: May want additional clarification here as the `set_` action's will overwrite the field with whatever `value` is provided. This aligns with `Set value as tags in rules` but not `Set index patterns in rules` -- may want to align all three?

> | `set_tags` | String[] |  Set value as tags in rules

Author: Agree with @spong . I rewritten it as

- Overwrite tags with value in rules
- Overwrite index patterns with value in rules

| `add_index` | String[] |  Adds index patterns to rules
| `delete_index` | String[] |  Delete index patterns from rules
| `set_index` | String[] |  Set index patterns in rules
| `set_timeline` a| object with 2 properties:

> * `timeline_id`: String

Reviewer 3: @vitaliidm or @banderror can you answer this? ^

> * `timeline_id`: String

Reviewer: So these two fields are actually part of the existing Rule CRUD API, just currently undocumented (they could be listed in the docs for [optional fields for all rule types](https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all), but we should verify as to why they weren't included initially). For the existing CRUD API these fields are optional ([source](https://github.com/elastic/kibana/blob/ba0833f42b89a0b62bccd1d8c04bf9c799f32258/x-pack/plugins/security_solution/common/detection_engine/schemas/request/rule_schemas.ts#L153-L154)), and so providing _either_ a `timeline_id` or `timeline_title` should validate.

As for the bulk edit API above though, it appears as if both are required ([source](https://github.com/elastic/kibana/blob/1f3f1c6fc6f8b4b8036c92161854b10814758708/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L462-L465)). 

For context, these are the same `id`/`name` fields used within the [Timeline API](https://www.elastic.co/guide/en/security/current/_get_timeline_template_by_templatetimelineid.html). As far as I can remember these fields are only being used for the `Investigate in Timeline` action on Alerts, and results in that specific alert data being opened in the specified timeline template (by means of [timelineId](https://github.com/elastic/kibana/blob/e312c36e4cf67dad48a58fdcd947eab022934315/x-pack/plugins/security_solution/public/detections/components/alerts_table/actions.tsx#L520)).

Hopefully that provides some context here, but let me know if you have anymore questions! We should discuss as a team if these fields can be documented/supported for the Rules API as a whole @banderror.

> * `timeline_id`: String

Reviewer: I forgot we don't expose a `Bulk Update Timeline` UI, so with these fields not being documented in the other Rule API's perhaps this doesn't need to be documented here just yet either?
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/2946766/156091255-563ee19a-61ff-48ef-8acf-23af7a659d4c.png" />
</p>

> * `timeline_id`: String

Author: Sounds reasonable to me.
I'm removing timeline parameter for now, until we add UI functionality

* `timeline_title`: String

>  |  Adds timeline to rule

Reviewer 4: ```suggestion
 |  Adds Timeline to rule
```

|==============================================

> Actions(<<bulk-edit-object-schema, BulkEditAction[]>> ) in `edit` array payload's property are applying in a historical order(from first to last).

Reviewer 4: ```suggestion
<<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the`edit` array payload's property.
```

[discrete]
===== Example requests

*Example 1*

> The following request will activate all rules that have the `test` tag:

Reviewer 4: ```suggestion
The following request activates all rules with the `test` tag:
```

to apply

.Possible values
* `enable`
* `disable`
* `delete`
* `duplicate`
* `export`
* `edit`

| Yes
| `edit` | <<bulk-edit-object-schema, BulkEditAction[]>>
| Edit object that describes applying an update action

|No.

Yes, if action is `edit`
|==============================================

[[bulk-edit-object-schema]]
[discrete]
==== BulkEditAction object

* `type` - enum - that defines what will be updated in rules
* `value` - any - value which will be applied in edit action.

[discrete]
===== Possible `BulkEditAction` object values

[width="100%",options="header"]
|==============================================
|`type` field |`value` field |Description
| `add_tags` | String[] |  Adds tags to rules
| `delete_tags` | String[] |  Delete tags from rules
| `set_tags` | String[] |  Set value as tags in rules
| `add_index` | String[] |  Adds index patterns to rules
| `delete_index` | String[] |  Delete index patterns from rules
| `set_index` | String[] |  Set index patterns in rules
| `set_timeline` a| object with 2 properties:

* `timeline_id`: String
* `timeline_title`: String

 |  Adds Timeline to rule
|==============================================

> <<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the`edit` array payload's property.

Reviewer 2: Missing space
```suggestion
<<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the `edit` array payload's property.
```

[discrete]
===== Response payload

> For `enable`, `disable`, `delete`, `edit`, and `duplicate` actions, a JSON object containing the outcome of the action: number of failed, succeeded, total rules, as well as full rule's objects(either created, deleted or updated depends on action) in array.

Reviewer 5: ```suggestion
For `enable`, `disable`, `delete`, `edit`, and `duplicate` actions, a JSON object containing the action's outcome: number of failed and successful outcomes, number of rules, and number of rule objects created, deleted or updated (depending on the action).
```

               "from":"now-45m",
               "rule_id":"7289bf08-4e91-4c70-bf01-e04c4c5d7756",
               "max_signals":100,
               "risk_score_mapping":[

               ],
               "severity_mapping":[

               ],
               "threat":[

               ],
               "to":"now",
               "references":[
                  "https://www.elastic.co/guide/en/security/current/prebuilt-ml-jobs.html"
               ],
               "version":6,
               "exceptions_list":[

               ],
               "immutable":false,
               "type":"machine_learning",
               "anomaly_threshold":50,
               "machine_learning_job_id":[
                  "packetbeat_dns_tunneling"
               ],
               "throttle":"no_actions",
               "actions":[

               ]
            }
         ],
         "created":[

         ],
         "deleted":[

         ]
      },
      "summary":{
         "failed":0,
         "succeeded":1,
         "total":1
      }
   }
}
--------------------------------------------------

For an `export` action, an `.ndjson` file containing exported rules.

*Example 2, Partial failure*

> The following request will add tags `tag-1`, `tag-2` to rules, which ids are send in payload:

Reviewer 5: ```suggestion
The following request adds tags `tag-1` and `tag-2` to the rules that have the IDs sent in the payload:
```

[source,console]
--------------------------------------------------
POST api/detection_engine/rules/_bulk_action
{
  "ids":[
    "8bc7dad0-9320-11ec-9265-8b772383a08d",
    "8e5c1a40-9320-11ec-9265-8b772383a08d",
  ],
  "action": "edit",
  "edit": [{ action: "add_tags", value:["tag-1", "tag-2"] }]
}
--------------------------------------------------

[discrete]
===== Response code

`500`::

>     Partial failure of bulk action

Reviewer 5: ```suggestion
    Indicates partial bulk action failure.
```

[discrete]
===== Response payload

> If processing of any rule fails, partial error will be return with id/name of affected rule and corresponding error together with successfully processed rules(same format as for successful 200 request)

Reviewer 5: ```suggestion
If processing of any rule fails, a partial error outputs the ID and/or name of the affected rule and the corresponding error, as well as successfully processed rules (in the same format as a successful 200 request).
```
